src/main/java/minijava/lexer/BasicLexerInput.java

```diff
@@ -0,0 +1,60 @@
+package minijava.lexer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import minijava.MJError;
+
+/** A basic lexer input stream without advanced buffering. */
+public class BasicLexerInput implements LexerInput {
+
+  private final InputStream stream;
+  private int currentChar = -2;
+  private int currentLine = 1;
+  private int currentColumn = 0;
+
+  public BasicLexerInput(InputStream stream) {
+    this.stream = stream;
+  }
+
+  @Override
+  public boolean hasNext() {
+    return currentChar != -1;
+  }
+
+  @Override
+  public Integer next() {
+    try {
+      currentChar = stream.read();
+      if (currentChar == '\n') {
```

Comment:
This is probably simplified too much (what about, for example, `\r`?), but since error locations are only bonus I'm okay with it.